### PR TITLE
Purchase optimization

### PR DESCRIPTION
**Pull request by Spirtix**
The idea is to aggregate the items to avoid individual processing, so the inventory, coins, and gems in the database are updated just once, which significantly improves performance.

### CODE REVIEW

File: src/Controllers/Common/ContentController.cs [hunk 1]
```diff
@@ -2293,4 +2194,106 @@ public IActionResult GetRevealIndex()
...
+        foreach (var i in itemsToSendBack) {
+            items.AddRange(Enumerable.Repeat(
+                new CommonInventoryResponseItem {
+                    CommonInventoryID = addedItems.ContainsKey(i.Key) ? addedItems[i.Key] : 0, // return inventory id if this item was added to the DB
+                    ItemID = i.Key,
+                    Quantity = 0
+                }, i.Value));
+        }
+        // NOTE: The quantity of purchased items is always 0 and the items are instead duplicated in both the request and the response.
+        //       Call AddItemToInventoryAndGetResponse with Quantity == 1 we get response with quantity == 0.
```
Comment (rpaciorek) [inline src/Controllers/Common/ContentController.cs:2291]: this note need be updated ... no more `AddItemToInventoryAndGetResponse` here but still we are need to use not obvious `Quantity = 0` and repeat `CommonInventoryResponseItem` for the same item ...

File: src/Controllers/Common/ContentController.cs [hunk 2]
```diff
@@ -2293,4 +2194,106 @@ public IActionResult GetRevealIndex()
             return adjectives[rand.Next(adjectives.Length)] + name;
         return name;
     }
+
+    private CommonInventoryResponse PurchaseItemsImpl(Viking viking, Dictionary<int, int> itemsToPurchase, bool mysteryBox) {
```
Comment (rpaciorek) [inline src/Controllers/Common/ContentController.cs:2198]: maybe `addAsMysteryBox` instead of just `mysteryBox`

File: src/Services/ItemService.cs [hunk 3]
```diff
@@ -42,13 +42,21 @@ public class ItemService {
...
 
         public int GetItemQuantity(ItemDataRelationship itemData) {
+            return GetItemQuantityImpl(itemData);
+        }
+
+        public int GetBulkItemQuantity(ItemDataRelationship itemData, int quantity) {
+            return GetItemQuantityImpl(itemData, quantity);
+        }
+
+        private int GetItemQuantityImpl(ItemDataRelationship itemData, int repeat = 1) {
```
Comment (rpaciorek) [inline src/Services/ItemService.cs:52]: maybe just `GetItemQuantity(ItemDataRelationship itemData, int bulkSize = 1)` instead of 3 functions

**Review by rpaciorek — COMMENTED** [review]
looks good ... only based on code - not tested by me (atm)

File: src/Controllers/Common/ContentController.cs [hunk 4]
```diff
@@ -2293,4 +2194,106 @@ public IActionResult GetRevealIndex()
...
+            return new CommonInventoryResponse {
+                Success = false,
+                CommonInventoryIDs = new CommonInventoryResponseItem[0],
+                UserGameCurrency = achievementService.GetUserCurrency(viking)
+            };
+        }
+
+        // deduct the cost of the purchase
+        achievementService.AddAchievementPoints(viking, AchievementPointTypes.GameCurrency, -totalCoinCost + gemsToAdd);
+        achievementService.AddAchievementPoints(viking, AchievementPointTypes.CashCurrency, -totalGemCost + coinsToAdd);
```
Comment (rpaciorek) [inline src/Controllers/Common/ContentController.cs:2275]: gems and coins are mixing here ... so when buy coins I get gems and vice versa

**Review by rpaciorek — COMMENTED** [review]
I tested this against 3.31, 1.13 and WoJS ... it seems to work OK, expect when purchasing gems/coins - they are swapped (when you buy gems, you get coins)

File: src/Controllers/Common/ContentController.cs (hunk 4 above)
Comment (Spirtix) [reply src/Controllers/Common/ContentController.cs]: Fixed

File: src/Controllers/Common/ContentController.cs (hunk 2 above)
Comment (Spirtix) [reply src/Controllers/Common/ContentController.cs]: Fixed

File: src/Controllers/Common/ContentController.cs (hunk 1 above)
Comment (Spirtix) [reply src/Controllers/Common/ContentController.cs]: Adjusted

File: src/Services/ItemService.cs (hunk 3 above)
Comment (Spirtix) [reply src/Services/ItemService.cs]: Adjusted

**Pull request merged**